### PR TITLE
10579: Bot protection: Fix all auth strategies

### PR DESCRIPTION
**Pull request by jlledom**
**What this PR does / why we need it**:

https://github.com/3scale/porta/pull/3659 broke OAuth2 and SSO Token login strategies. This fixes the regression.

**Which issue(s) this PR fixes** 

[THREESCALE-10579](https://issues.redhat.com/browse/THREESCALE-10579)

**Notes**

AFAIK, there are 4 auth strategies for the Developer Portal

- `internal`: Provide username + password from the login screen
- `oauth2`: Using external authentication providers configured from the "Single Sign-on Integrations" form at `/p/admin/authentication_providers`
- `sso`: Using a generated auth token, as documented here: https://access.redhat.com/documentation/en-us/red_hat_3scale_api_management/2.14/html-single/creating_the_developer_portal/index#sso
- `cas`: Some old and probably unused auth method, I don't know about it.

Note how confusing is we internally call `oauth2` to something we call "Single Sign-on Integrations" in the UI; and at the same time we internally call `sso` to something which has nothing to do with the "SSO Integrations" screen or SSO at all.

The commit that actually fixes the bug is https://github.com/3scale/porta/pull/3669/commits/859aff4e223e93fa227581c87412beee4b9f174a. It disables the bot check for all auth strategies except `internal`.

I also refactored the cucumbers to ensure we have integration tests for `internal`, `oauth` and `sso`. In commits https://github.com/3scale/porta/pull/3669/commits/5c23b5d169e4ccfbf63d54d30af26e0a5cb82cd0 and https://github.com/3scale/porta/pull/3669/commits/40367f68984c6dec37ed5f4cb69a044b98a7c7f5

**Verification steps**

- Internal: Follow the instructions at https://github.com/3scale/porta/pull/3659. It should work the same way
- OAuth2: Configure any OAuth ID providers, for instance GitHub, as explained here: https://access.redhat.com/documentation/en-us/red_hat_3scale_api_management/2.14/html-single/creating_the_developer_portal/index#auth-github
  - When the process requires a new Sign Up, that should work as well.
- SSO Token: Foolow the instructions here: https://access.redhat.com/documentation/en-us/red_hat_3scale_api_management/2.14/html-single/creating_the_developer_portal/index#sso

### CODE REVIEW

File: app/lib/three_scale/bot_protection/base.rb [hunk 1]
```diff
@@ -3,12 +3,20 @@
...
       def bot_protection_level
         site_account.settings.spam_protection_level
       end
 
       def bot_protection_enabled?
-        Recaptcha.captcha_configured? && bot_protection_level != :none
+        Recaptcha.captcha_configured? && bot_protection_level != :none && !request_from_auth_provider?
+      end
+
+      def request_from_auth_provider?
```
Comment (akostadinov) [inline app/lib/three_scale/bot_protection/base.rb:16]: is this a valid check for all classes that include this module?

Comment (jlledom) [reply app/lib/three_scale/bot_protection/base.rb]: That's old code. You reviewed too fast!

File: features/step_definitions/authentication/sso_token_steps.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+Given "the provider generated a sso token for the buyer" do
+  @sso_token = FactoryBot.build(:sso_token, account: @provider)
+  @sso_token.save
+end
+
+Given "another provider generated a sso token for another buyer" do
+  buyer = FactoryBot.create(:buyer_account)
+  @sso_token = FactoryBot.build(:sso_token, account: buyer.provider_account, user_id: buyer.users.first.id)
+  @sso_token.save
```
Comment (mayorova) [inline features/step_definitions/authentication/sso_token_steps.rb:11]: Couldn't you just do `.create` and avoid `.save` ?

File: features/step_definitions/provider_steps.rb [hunk 3]
```diff
@@ -231,7 +231,15 @@ def setup_provider(login)
 end
 
 When(/I authenticate by Oauth2$/) do
-  try_login_with_sso
+  try_buyer_login_oauth
```
Comment (mayorova) [inline features/step_definitions/provider_steps.rb:234]: Hm... why are buyer login steps defined under `provider_steps`? Am I missing something?

File: features/step_definitions/provider_steps.rb [hunk 4]
```diff
@@ -124,7 +124,7 @@ def login_form
   GHERKIN
 end
 
-And('As a developer, I login through RH SSO') do
+And('As a developer, I want to login through RH SSO') do
```
Comment (mayorova) [inline features/step_definitions/provider_steps.rb:127]: Why this change? I'd keep "I log in" (well, actually "log in" is two words :grimacing: )

File: features/step_definitions/authentication/sso_token_steps.rb (hunk 2 above)
Comment (jlledom) [reply features/step_definitions/authentication/sso_token_steps.rb:11]: No, because `Factorybot.create` calls `#save!` which is not implemented by `SSOToken` as it doesn't inherit from `ActiveRecord`.

Comment (mayorova) [reply features/step_definitions/authentication/sso_token_steps.rb:11]: OK, thanks for the clarification :ok_hand:

File: features/developer_portal/authentication/sso_token.feature [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    And a buyer "bob" signed up to provider "foo.3scale.localhost"
+    And the current domain is foo.3scale.localhost
+
+  Scenario: Buyer can't login with an invalid token
+    Given another provider generated a sso token for another buyer
+    When the buyer authenticates by SSO Token
+    Then the page should contain "Invalid SSO Token"
+
+  Scenario: Buyer can't login with an expired token
+    Given the provider generated a sso token for the buyer
```
Comment (mayorova) [inline features/developer_portal/authentication/sso_token.feature:17]: Just an improvement suggestion - the step could potentially have an optional suffix "valid for N minutes", that would create a token with `expires_in == N` (if missing, fall back to default 10 minutes).

It would make it clearer in this scenario that if the token is valid for 10 minutes, and 15 pass - that it would not be valid anymore.
But it's OK if you think that's too much.

File: features/step_definitions/provider_steps.rb (hunk 4 above)
Comment (jlledom) [reply features/step_definitions/provider_steps.rb:127]: I think the step name was wrong because it doesn't login, it only opens the login screen.

File: features/developer_portal/authentication/sso_token.feature (hunk 5 above)
Comment (jlledom) [reply features/developer_portal/authentication/sso_token.feature:17]: That's fine, I'll do it 👍

File: features/step_definitions/buyer_steps.rb [hunk 6]
```diff
@@ -138,6 +138,22 @@ def login_form
...
+When "the developer tries to log in" do
+  username = @buyer.admins.first.username
+  set_current_domain("foo.3scale.localhost")
+  visit path_to('the login page')
+  fill_in('Username or Email', with:  username)
+  fill_in('Password', with: "supersecret")
+  click_button('Sign in')
+end
+
+Then "the developer login attempt failed" do
```
Comment (mayorova) [inline features/step_definitions/buyer_steps.rb:150]: 1. I'd say "login attempt fails", "failed" reads weird in a scenario
2. Shouldn't we be more specific, and mention recaptcha in the step, and also verify the recaptcha error in flash message?

File: features/step_definitions/provider_steps.rb (hunk 4 above)
Comment (mayorova) [reply features/step_definitions/provider_steps.rb:127]: Ah, OK... 

Then I think to make it more clear this step should be "I see RH-SSO login option on the login page".
But I don't know if it's good either, for sure @josemigallas  wouldn't like it :grimacing:

**Review by mayorova — APPROVED** [review]
LGTM :+1:

File: features/step_definitions/provider_steps.rb (hunk 4 above)
Comment (jlledom) [reply features/step_definitions/provider_steps.rb]: done

File: features/developer_portal/authentication/sso_token.feature (hunk 5 above)
Comment (jlledom) [reply features/developer_portal/authentication/sso_token.feature]: done

File: features/step_definitions/buyer_steps.rb (hunk 6 above)
Comment (jlledom) [reply features/step_definitions/buyer_steps.rb]: I renamed the step, but I couldn't make it more specific. I was trying to find the recaptcha error in the resulting screen, but the error doesn't appear because the recaptcha method which adds the error is stubbed. I detect the login failure by checking we are in the login screen again but having `/session` as URL. AFAIK, the only way to see the login screen under `/session` is after a failed login attempt. I hope that's good enough.

File: features/step_definitions/authentication/sso_token_steps.rb (hunk 2 above)
Comment (akostadinov) [reply features/step_definitions/authentication/sso_token_steps.rb:18]: Then is it appropriate to use FactoryBot to begin with? I thought it is supposed to be only used for models. Or am I mistaken?

Comment (jlledom) [reply features/step_definitions/authentication/sso_token_steps.rb:18]: There isn't a clear answer, `sso_token.rb` is under `/models` and implements some `ActiveRecord` features like validation, but it's not based on a DB table. So it's somewhere between a plain object and a Rails model. It seems to me that using FactoryBot or not is equally correct.

**Review by mayorova — APPROVED** [review]
